pkg/paho-mqtt/contrib/riot_iface.c

```diff
@@ -57,20 +57,23 @@ static int mqtt_read(struct Network *n, unsigned char *buf, int len,
     void *_buf;
     int rc = -1;
 
-    if (IS_USED(MODULE_LWIP)) {
-        /* As LWIP doesn't support packet reading byte per byte and
-         * PAHO MQTT reads like that to decode it on the fly,
-         * we read TSRB_MAX_SIZE at once and keep them in a ring buffer.
-         */
-        _buf = _temp_buf;
-        _len = TSRB_MAX_SIZE;
-        _timeout = 0;
-    }
-    else {
-        _buf = buf;
-        _len = len;
-        _timeout = timeout_ms;
-    }
+#if defined(MODULE_LWIP)
+    /* As LWIP doesn't support packet reading byte per byte and
+     * PAHO MQTT reads like that to decode it on the fly,
+     * we read TSRB_MAX_SIZE at once and keep them in a ring buffer.
+     */
+    _buf = _temp_buf;
+    _len = TSRB_MAX_SIZE;
+    _timeout = 0;
+#elif defined(MODULE_GNRC)
+    _buf = buf;
+    _len = len;
+    _timeout = 0;
+#else
+    _buf = buf;
+    _len = len;
+    _timeout = timeout_ms;
+#endif
```

Comment:
Would it compile if the conditionals are still left at C level?

Note that the `IS_USED()` macro resolves to the numeric literal `0` or `1` at preprocessor time. Hence, it is trivial for the compiler to optimize out dead branches. Only with `-O0` it would be actually be more efficient to use the preprocessor, in all other cases there is no cost. You can even call functions that are declared by not implemented (e.g. if the module `tsrb` is not used with GNRC but only called in the `IS_USED(MODULE_LWIP)` case, there will be no undefined symbol error at link time, since the function calls in the dead branches are optimized out).

The advantage of keeping it a C level is that the C compiler still performs syntax checks and diagnostics on all variants and issues are easier found. For that reason, we generally prefer C level conditionals over preprocessor conditionals.

Note that accessing e.g. `struct` members that are only provided when a given module is used will not work without the preprocessors, as even in dead branches C code may not accesses non-existing `struct` members. So there cases in which avoiding the preprocessor is not possible (or would require carrying unused `struct` members into all configurations).

Reply:
It is possible to replace the `#if defined(...)` with `if (IS_USED(...))` below in [pkg/paho-mqtt/contrib/riot_iface.c#L141](https://github.com/RIOT-OS/RIOT/pull/17004/commits/3ceb4154f6c1de2b778bf7b03b18d6fcce6f9d53#diff-4da45cb33725fb03c458188e7e6f8eff55a3d78a3bbe8ea7de1430c74f87b200L141), so I have done that.

In all the other cases, replacing it causes compiler errors since the code segments use the variables `buffer`, `_temp_buf` and/or `tsrb_lwip_tcp` which are not declared if LWIP isn't used ([pkg/paho-mqtt/contrib/riot_iface.c#L42-L46](https://github.com/RIOT-OS/RIOT/blob/3ceb4154f6c1de2b778bf7b03b18d6fcce6f9d53/pkg/paho-mqtt/contrib/riot_iface.c#L42-L46)).

Reply:
It should be possible to just declare the variables in all cases. Both GCC and clang will just garbage collect unused variable, if they are used only in known-to-be dead branches.

Reply:
OK, that also works :) I have made the changes.